lib/metrics.js

```diff
@@ -0,0 +1,134 @@
+const {
+  Registry,
+  Counter,
+  Histogram,
+  Gauge,
+  collectDefaultMetrics,
+} = require('prom-client');
+const { cpu, mem, netstat } = require('node-os-utils');
+
+const METRICS_URL = '/metrics';
+const metricsURL = process.env.METRICS_URL || METRICS_URL;
+
+module.exports = function(config) {
+  // eslint-disable-next-line max-len
+  // See: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/semantic_conventions/faas.md
+  // TODO: label names should use . instead of _ as a word separator,
+  // but the prom-client module throws an InvalidLabelName exception.
+  const labels = {
+    'faas_name': config.name || 'anonymous',
+    'faas_id': config.id || '',
+    'faas_instance': config.instance || '',
+    'faas_runtime': 'Node.js',
+  };
+
+  // Create a metrics register and add our identifying labels to it
+  const register = new Registry();
+  register.setDefaultLabels(labels);
+  collectDefaultMetrics({ register });
+
+  // invoked on GET to /metrics
+  async function callMetrics(req, res) {
+    try {
+      res.headers('Content-Type', register.contentType);
+      res.send(await register.metrics());
+    } catch (ex) {
+      res.code(500).send(ex);
+    }
+  }
+
+  const invocations = new Counter({
+    name: 'faas_invocations',
+    help: 'The number of times this function was invoked',
+    registers: [register],
+  });
+
+  const execLatency = new Histogram({
+    name: 'faas_execution_latency',
+    help: 'The time it took for this function to be invoked and return',
+    registers: [register],
+  });
+
+  const errors = new Counter({
+    name: 'faas_errors',
+    help: 'The number of function invocation failures',
+    registers: [register],
+  });
+
+  const coldStartLatency = new Histogram({
+    name: 'faas_cold_start_latency',
+    help: 'The time it took for this function to be in the ready state after the container started',
+    registers: [register],
+  });
+
+  const queueLatency = new Histogram({
+    name: 'faas_queue_latency',
+    help: 'The time spent in a system queue before the function was invoked',
+    registers: [register],
+  });
+
+  new Gauge({
+    name: 'faas_cpu_utilization',
+    help: 'The CPU utilization of this function instance',
+    collect() {
+      cpu.usage()
+        .then(usage => this.set(usage));
+    },
+    registers: [register],
+  });
+
+  new Gauge({
+    name: 'faas_mem_utilization',
+    help: 'The memory utilization of this function instance',
+    collect() {
+      mem.used()
+        .then(used => this.set(used.usedMemMb));
+    },
+    registers: [register],
+  });
+
+  new Gauge({
+    name: 'faas_netio_utilization',
+    help: 'The network I/O utilization of this function instance',
+    collect() {
+      netstat.inOut()
+        .then(io => this.set(io.total.inputMb + io.total.outputMb));
```

Comment:
The `io` is a string `"not supported"` on my machine I think that should be handled.

Reply:
```diff
diff --git a/lib/metrics.js b/lib/metrics.js
index 2ea7c59..facd7a5 100644
--- a/lib/metrics.js
+++ b/lib/metrics.js
@@ -70,9 +70,9 @@ module.exports = function(config) {
   new Gauge({
     name: 'faas_cpu_utilization',
     help: 'The CPU utilization of this function instance',
-    collect() {
-      cpu.usage()
-        .then(usage => this.set(usage));
+    async collect() {
+      const usage = await cpu.usage();
+      this.set(usage);
     },
     registers: [register],
   });
@@ -80,9 +80,9 @@ module.exports = function(config) {
   new Gauge({
     name: 'faas_mem_utilization',
     help: 'The memory utilization of this function instance',
-    collect() {
-      mem.used()
-        .then(used => this.set(used.usedMemMb));
+    async collect() {
+      const used = await mem.used();
+      this.set(used.usedMemMb);
     },
     registers: [register],
   });
@@ -90,9 +90,11 @@ module.exports = function(config) {
   new Gauge({
     name: 'faas_netio_utilization',
     help: 'The network I/O utilization of this function instance',
-    collect() {
-      netstat.inOut()
-        .then(io => this.set(io.total.inputMb + io.total.outputMb));
+    async collect() {
+      const io = await netstat.inOut();
+      if (typeof io !== 'string') {
+        this.set(io.total.inputMb + io.total.outputMb);
+      }
     },
     registers: [register],
     labels,
```

Reply:
Good catch and thanks for the suggestion. I've done it a little bit differently, using `isNotSupported()` from the `node-os-utils` module.